File: src/nodding_lidar/nodding_lidar/oscillation_controller.py
#!/usr/bin/env python3
"""
oscillation_controller.py - Controls servo that nods RPLidar to convert 2D to 3D
For NIDAR AirMouse: RPLidar A2/A3 mounted on servo (e.g., Dynamixel XL430 or MG996R via Arduino)

Mechanism:
- Servo oscillates pitch from -45° to +45° (or -30° to +60° to cover floor to ceiling)
- At each angle, 2D scan (360° horizontal) is captured
- Combined to form 3D pointcloud
- Frequency: 0.5-1 Hz nodding, 10Hz lidar = ~20 scans per sweep

Hardware options:
1. Arduino + Servo: Arduino subscribes to /servo_angle_cmd, publishes /servo_angle
2. Dynamixel SDK: Direct ROS2 control via dynamixel_sdk
3. PWM via PCA9685 on Jetson/RPi

This node publishes servo angle and controls oscillation.
"""
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64, Bool
from sensor_msgs.msg import JointState
import math
import time

class OscillationController(Node):
    def __init__(self):
        super().__init__('oscillation_controller')
        self.declare_parameter('min_angle_deg', -45.0)  # min pitch
        self.declare_parameter('max_angle_deg', 45.0)   # max pitch
        self.declare_parameter('frequency_hz', 0.5)     # oscillation freq
        self.declare_parameter('mode', 'sinusoidal')    # sinusoidal or triangular
        self.declare_parameter('servo_topic', '/servo_angle_cmd')
        self.declare_parameter('enable', True)

        self.min_angle = math.radians(self.get_parameter('min_angle_deg').value)
        self.max_angle = math.radians(self.get_parameter('max_angle_deg').value)
        self.freq = self.get_parameter('frequency_hz').value
        self.mode = self.get_parameter('mode').value
        self.servo_topic = self.get_parameter('servo_topic').value

        # Publishers
        self.angle_pub = self.create_publisher(Float64, self.servo_topic, 10)
        self.joint_pub = self.create_publisher(JointState, '/joint_states', 10)
        self.scan_enable_pub = self.create_publisher(Bool, '/nodding_scan_enable', 10)

        # State
        self.start_time = time.time()
        self.current_angle = 0.0

        # Timer 50Hz for smooth servo
        self.timer = self.create_timer(0.02, self.update)

        self.get_logger().info(f"Nodding Lidar Controller: {self.get_parameter('min_angle_deg').value}° to {self.get_parameter('max_angle_deg').value}° @ {self.freq}Hz, mode {self.mode}")

    def update(self):
        elapsed = time.time() - self.start_time

        # Compute target angle
        if self.mode == 'sinusoidal':
            # Sinusoidal oscillation: smooth
            amplitude = (self.max_angle - self.min_angle) / 2
            center = (self.max_angle + self.min_angle) / 2
            self.current_angle = center + amplitude * math.sin(2 * math.pi * self.freq * elapsed)
        else:
            # Triangular (sawtooth): linear sweep
            period = 1.0 / self.freq
            phase = (elapsed % period) / period
            if phase < 0.5:
                # Up sweep
                self.current_angle = self.min_angle + (self.max_angle - self.min_angle) * (phase * 2)
            else:
                # Down sweep
                self.current_angle = self.max_angle - (self.max_angle - self.min_angle) * ((phase - 0.5) * 2)

        # Publish angle command
        msg = Float64()
        msg.data = self.current_angle
        self.angle_pub.publish(msg)

        # Publish joint state for TF and 3D reconstruction
        joint_msg = JointState()
        joint_msg.header.stamp = self.get_clock().now().to_msg()
        joint_msg.name = ['rplidar_pitch_joint']
        joint_msg.position = [self.current_angle]
        joint_msg.velocity = [0.0]
        self.joint_pub.publish(joint_msg)

        # Enable scanning always (or disable at extremes to reduce distortion)
        enable_msg = Bool()
        enable_msg.data = True
        self.scan_enable_pub.publish(enable_msg)

def main(args=None):
    rclpy.init(args=args)
    node = OscillationController()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/nodding_lidar_cpp/src/oscillation_controller.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float64.hpp>
#include <std_msgs/msg/bool.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <cmath>

class OscillationController : public rclcpp::Node {
public:
  OscillationController() : Node("oscillation_controller") {
    this->declare_parameter("min_angle_deg", -45.0);
    this->declare_parameter("max_angle_deg", 45.0);
    this->declare_parameter("frequency_hz", 0.5);
    this->declare_parameter("mode", "sinusoidal");

    double min_deg = this->get_parameter("min_angle_deg").as_double();
    double max_deg = this->get_parameter("max_angle_deg").as_double();
    freq_ = this->get_parameter("frequency_hz").as_double();
    mode_ = this->get_parameter("mode").as_string();

    min_angle_ = min_deg * M_PI/180.0;
    max_angle_ = max_deg * M_PI/180.0;

    angle_pub_ = this->create_publisher<std_msgs::msg::Float64>("/servo_angle_cmd", 10);
    joint_pub_ = this->create_publisher<sensor_msgs::msg::JointState>("/joint_states", 10);
    enable_pub_ = this->create_publisher<std_msgs::msg::Bool>("/nodding_scan_enable", 10);

    start_time_ = this->now();
    timer_ = this->create_wall_timer(std::chrono::milliseconds(20), std::bind(&OscillationController::update, this));

    RCLCPP_INFO(this->get_logger(), "C++ Nodding Controller %.1f to %.1f deg @ %.2f Hz", min_deg, max_deg, freq_);
  }

private:
  void update() {
    double elapsed = (this->now() - start_time_).seconds();
    double angle;
    if (mode_=="sinusoidal") {
      double amp = (max_angle_ - min_angle_)/2;
      double center = (max_angle_ + min_angle_)/2;
      angle = center + amp * std::sin(2*M_PI*freq_*elapsed);
    } else {
      double period = 1.0/freq_;
      double phase = std::fmod(elapsed, period)/period;
      if (phase<0.5) angle = min_angle_ + (max_angle_-min_angle_)*(phase*2);
      else angle = max_angle_ - (max_angle_-min_angle_)*((phase-0.5)*2);
    }

    std_msgs::msg::Float64 msg; msg.data=angle; angle_pub_->publish(msg);

    sensor_msgs::msg::JointState js;
    js.header.stamp=this->now();
    js.name={"rplidar_pitch_joint"};
    js.position={angle};
    js.velocity={0};
    joint_pub_->publish(js);

    std_msgs::msg::Bool en; en.data=true; enable_pub_->publish(en);
  }

  double min_angle_, max_angle_, freq_;
  std::string mode_;
  rclcpp::Time start_time_;
  rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr angle_pub_;
  rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr joint_pub_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr enable_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<OscillationController>());
  rclcpp::shutdown();
  return 0;
}
